kevlar/include/kevlar_bits/model/binomial/berry_inla.hpp

```diff
@@ -0,0 +1,369 @@
+
+    const size_t n_arm_samples;
+    const vec_t efficacy_thresholds;
+
+    // variables dependent solely on the quadrature points (independent of the
+    // data).
+    const vec_t quad_wts;
+    const mat_t cov;
+    const mat_t neg_precQ;
+    const vec_t logprecQdet;
+    const vec_t logprior;
+    const value_t opt_tol2;
+    const value_t logit_p1;
+
+   public:
+    template <class _GenType, class _ValueType, class _UIntType,
+              class _GridRangeType>
+    struct SimGlobalState;
+
+    template <class _GenType, class _ValueType, class _UIntType,
+              class _GridRangeType>
+    using sim_global_state_t =
+        SimGlobalState<_GenType, _ValueType, _UIntType, _GridRangeType>;
+
+    template <class _GridRangeType>
+    using kevlar_bound_state_t = KevlarBoundStateFixedNDefault<_GridRangeType>;
+
+    BerryINLA(size_t n_arm_samples, const Eigen::Ref<const vec_t>& cv,
+              const Eigen::Ref<const vec_t>& efficacy_thresholds,
+              const Eigen::Ref<const vec_t>& quad_wts,
+              const Eigen::Ref<const mat_t>& cov,
+              const Eigen::Ref<const mat_t>& neg_precQ,
+              const Eigen::Ref<const vec_t>& logprecQdet,
+              const Eigen::Ref<const vec_t>& logprior, value_t opt_tol,
+              value_t logit_p1)
+        : arm_base_t(ARMS, n_arm_samples),
+          base_t(),
+          n_arm_samples(n_arm_samples),
+          efficacy_thresholds(efficacy_thresholds),
+          quad_wts(quad_wts),
+          cov(cov),
+          neg_precQ(neg_precQ),
+          logprecQdet(logprecQdet),
+          logprior(logprior),
+          opt_tol2(opt_tol * opt_tol),
+          logit_p1(logit_p1) {
+        assert(efficacy_thresholds.size() == ARMS);
+        set_critical_values(cv);
+    }
+
+    using base_t::critical_values;
+    void set_critical_values(const Eigen::Ref<const colvec_type<value_t>>& cv) {
+        auto& cv_ = base_t::critical_values();
+        cv_ = cv;
+        std::sort(cv_.begin(), cv_.end(), std::greater<value_t>());
+    }
+
+    template <class _GenType, class _ValueType, class _UIntType,
+              class _GridRangeType>
+    auto make_sim_global_state(const _GridRangeType& grid_range) const {
+        return sim_global_state_t<_GenType, _ValueType, _UIntType,
+                                  _GridRangeType>(*this, grid_range);
+    }
+
+    template <class _GridRangeType>
+    auto make_kevlar_bound_state(const _GridRangeType& gr) const {
+        return kevlar_bound_state_t<_GridRangeType>(n_arm_samples, gr);
+    }
+
+    vec_t get_posterior_exceedance_probs(const vec_t& y) const {
+        assert((y.array() >= 0).all());
+        assert((y.array() <= n_arm_samples).all());
+        size_t nsig2 = quad_wts.size();
+
+        vec_t exceedances(ARMS);
+        exceedances.setZero();
+
+        double sigma2_integral = 0.0;
+        for (int s = 0; s < nsig2; s++) {
+            std::array<double, ARMS> t{};
```

Comment:
Is there a reason for using `array`? Why not `Eigen`? You can vectorize almost all of this, especially since your `ARMS` is >= 2.

Reply:
I copied it over from fast_inla_ext.cpp and I intend to delete this code pretty soon. So... basically I didn't want to waste effort on it! =) Do you mind leaving it like this for now?

Reply:
Amusingly eigen has a near-clone of std::array due to the dark days before C++11.